tardis/plasma/properties/nlte_rate_equation_solver.py

```diff
@@ -0,0 +1,367 @@
+            total_rad_recomb_coefficients,
+            total_coll_ion_coefficients,
+            total_coll_recomb_coefficients,
+        ) = self.prepare_ion_recomb_coefficients_nlte_ion(
+            gamma,
+            alpha_sp,
+            alpha_stim,
+            coll_ion_coeff,
+            coll_recomb_coeff,
+            partition_function,
+            levels,
+            level_boltzmann_factor,
+        )
+
+        # >>>TODO:initial electron density should be included in the initial guess, added in a future PR
+        initial_electron_density = number_density.sum(axis=0)
+        # <<<
+        rate_matrix = self.calculate_rate_matrix(
+            phi[0],
+            initial_electron_density[0],
+            rate_matrix_index,
+            total_photo_ion_coefficients[0],
+            total_rad_recomb_coefficients[0],
+            total_coll_ion_coefficients[0],
+            total_coll_recomb_coefficients[0],
+        )
+        return (
+            -1,
+            -1,
+        )  # function still empty, that's why return statement is arbitrary at this point
+
+    @staticmethod
+    def calculate_rate_matrix(
+        phi_shell,
+        electron_density,
+        rate_matrix_index,
+        total_photo_ion_coefficients,
+        total_rad_recomb_coefficients,
+        total_coll_ion_coefficients,
+        total_coll_recomb_coefficients,
+    ):
+        """
+
+        Parameters
+        ----------
+        phi_shell : DataFrame
+            Saha Factors in the current shell
+        electron_density : float
+            Guess for electron density in the current shell
+        rate_matrix_index : MultiIndex
+            Index used for constructing the rate matrix
+        total_photo_ion_coefficients : DataFrame
+            Photo ionization coefficients
+        total_rad_recomb_coefficients : DataFrame
+            Radiative recombination coefficients(should get multiplied by electron density)
+        total_coll_ion_coefficients : DataFrame
+            Collisionional ionization coefficients(should get multiplied by electron density)
+        total_coll_recomb_coefficients : DataFrame
+            Collisional recombination coefficients (should get multiplied by electron density^2)
+
+        Returns
+        -------
+        DataFrame
+            Rate matrix used for nlte solver.
+        """
+        rate_matrix = pd.DataFrame(
+            0.0, columns=rate_matrix_index, index=rate_matrix_index
+        )
+        total_rad_recomb_coefficients = (
+            total_rad_recomb_coefficients * electron_density
+        )
+        total_coll_ion_coefficients = (
+            total_coll_ion_coefficients * electron_density
+        )
+        total_coll_recomb_coefficients = (
+            total_coll_recomb_coefficients * electron_density**2
+        )
+        atomic_numbers = (
+            rate_matrix_index.get_level_values(0).unique().drop("n_e")
+        )  # dropping the n_e index
```

Comment:
Why?

Reply:
`rate_matrix_index`'s 1st index is atomic number and "n_e" added in the end for constructing the `rate_matrix` in the right shape. That's why I'm getting rid of the last index and dropping repeated values(as each atomic number has multiple corresponding rows)

Reply:
can you put this in the comments